HW1/P4/P4.py

```diff
@@ -0,0 +1,55 @@
+import findspark
+findspark.init()
+import pyspark
+
+# num partitions (4 times the number of cores)
+N = 16
+
+def bfs(root, graph):
+	# initialize queue and distances rdds
+	distances = queue = sc.parallelize([(root, 0)], N)
+	acc = sc.accumulator(1)
+	iteration = 0
+	while acc.value != 0:
+		acc = sc.accumulator(0)
+		iteration += 1
+		# keep a temporary copy of the queue to later on remove items in queue from graph
+		temp = queue
+		# get all (neighbor, distance) value pairs for each key in queue
+		joined = graph.join(queue, N).distinct()
+		# have to partition again when taking the values
+		queue = joined.values().distinct().partitionBy(N)
+		# take out original queue keys from graph and cache for later iterations
+		graph = graph.subtractByKey(temp, N).cache()
+		# set the distance values to current iteration value
+		newDistances = queue.mapValues(lambda _: iteration)
+		# remove already visited keys from the queue, and cache for later iterations
+		queue = newDistances.subtractByKey(distances, N).cache()
+		# add keys in queue with their distances to distances rdd, and cache for later iterations
+		distances = distances.union(queue).partitionBy(N).cache()
+		# increment accumulator to satisfy the while condition (unless the queue is empty, in which case the loop stops after this)
+		queue.foreach(lambda _: acc.add(1))
+	return (distances.count(), distances)
```

Comment:
Nice implementation, and good documentation.
The accumulator can be probably updated during the queue.mapValues invocation.